### PR TITLE
Elixir Module support

### PR DESCRIPTION
**Pull request by wingyplus**
Let's try:

```
$ ./hack/dev
$ export PATH=$PWD/hack:$PWD/bin:$PATH
$ with-dev dagger init --sdk=github.com/dagger/dagger/sdk/elixir/runtime potato
Initialized module potato in potato
$ cd potato
$ with-dev dagger functions
Name             Description
container-echo   -
grep-dir         -
$ with-dev dagger call container-echo --string-arg 'Hello, Dagger' stdout
Hello, Dagger
```

🎉 

## Technical design

CAUTION: This is the long road to make it done. The design can be change depends on the discussion and feedback.

Let's take a look at module that SDK generate for you.

```elixir
defmodule Potato do
  @moduledoc """
  A generated module for Potato functions.
  """

  use Dagger.Mod, name: "Potato"

  defstruct [:dag]

  @doc """
  Returns a container that echos whatever string argument is provided.
  """
  @function [
    args: [
      string_arg: [type: :string]
    ],
    return: Dagger.Container
  ]
  def container_echo(self, args) do
    self.dag
    |> Dagger.Client.container()
    |> Dagger.Container.from("alpine:latest")
    |> Dagger.Container.with_exec(~w"echo #{args.string_arg}")
  end

  @function [
    args: [
      directory_arg: [type: Dagger.Directory],
      pattern: [type: :string]
    ],
    return: :string
  ]
  def grep_dir(self, %{directory_arg: directory, pattern: pattern}) do
    self.dag
    |> Dagger.Client.container()
    |> Dagger.Container.from("alpine:latest")
    |> Dagger.Container.with_mounted_directory("/mnt", directory)
    |> Dagger.Container.with_workdir("/mnt")
    |> Dagger.Container.with_exec(["grep", "-R", pattern, "."])
    |> Dagger.Container.stdout()
  end
end
```

Any module that want to be a Dagger module needs to call `use Dagger.Mod` and need to declare the name of the module using `:name` option.

Each function has a module attribute called `function`. The value of that attribute is a keyword (for non-Elixir developer, think like a ordered key/value in another language) that contain the information of the function, currently has only argument and the return type. The module runtime will use this information to generate module declaration using `Module` API and do encode/decode value between Dagger and Elixir world.

The reason I declare this is because in Elixir, we have no other ways to obtain type from function argument because it is a dynamic programming language. But luckily that we can use data-structure to describe the function structure. :)

After declare the module, it's time to register a module. Take a look at `application.ex` inside `dagger/potato/lib/potato`

```elixir
defmodule Potato.Application do
  # See https://hexdocs.pm/elixir/Application.html
  # for more information on OTP Applications
  @moduledoc false

  use Application

  @impl true
  def start(_type, _args) do
    children = [
      Dagger.ModuleRuntime.Registry, 
      Potato
    ]

    # See https://hexdocs.pm/elixir/Supervisor.html
    # for other strategies and supported options
    opts = [strategy: :one_for_one, name: Potato.Supervisor]
    Supervisor.start_link(children, opts)
  end
end
```

In Elixir, they has a concept called `Supervisor` which's a process (the lightweight process like a goroutine) that maintain processes defined in `children` to keep alive.  We put it under `Application` and tell the Erlang/OTP to start this application when function start invoking. 

In children, we have 2 process, the first one is module registry that store Dagger module information to uses at runtime, and the second one is the Dagger module implementation, it's `Potato` in the example above.

## Tasks that need to done in this PR

- [x] Description support for module and function.
- [x] Support all primitive types. String already support but other types are not (integer, boolean, etc.).
- [x] Cleanup project code generator

## Tasks that can follow up after this PR

- [ ] Optional argument.
- [ ] ~Field support (like `@field` in TypeScript). I still doubt how to do this at the moment. :(~
- [ ] Improve the documentation in `Dagger.ModuleRuntime`.
- [ ] Decode object type back to struct.

### CODE REVIEW

File: sdk/elixir/lib/dagger_module_runtime.ex
Comment (wingyplus) [inline sdk/elixir/lib/dagger_module_runtime.ex:6]: To reviewer, the `__on_definition__` invoke when Elixir start compile the source. When it found a function declaration, it'll pass the information to this function.

**Comment by gerhard** [comment]
This looks amazing! WDYT @helderco @gmile?

What else is on your list @wingyplus before this is ready for the final review & merge?

**Comment by gmile** [comment]
Yes, I've been watching this PR for a while, this is really promising, thanks for ping @gerhard!

The kind of blocker I have with my article for Elixir SDK is entering a shell of the module to do interactive debugging inside, e.g. https://github.com/dagger/dagger/issues/4463 🤔 Is there anywhere I can read more about entering an interactive shell for a module? If not, is it because it's not possible yet in principle, or this kind of interactivity is not planned in the world of modules/functions?

To put it differently, let's say my dagger module runs `mix test` command, and it fails with a very, very weird error that's difficult to understand how to fix unless you're in a shell, looking around, poking files, etc. - interactively.

**Comment by wingyplus** [comment]
> What else is on your list @wingyplus before this is ready for the final review & merge?

@gerhard I have add a checklist on the PR description.  The checklist has some items that need to be done and some of it could follow-up later.  The follow-up tasks may also convert to the issue after this PR ready to review.

BTW, the Elixir module is now working, if anyone have time to review, I'm really appreciate. :)

**Comment by wingyplus** [comment]
The function doc is now supported like screenshot below.

Screenshot 2024-04-03 230144

If the plan is going well, this PR, the MVP version, can go out of draft version within next week. :)

**Comment by wingyplus** [comment]
Note for me:

- [ ] add the limitation to the description

**Comment by wingyplus** [comment]
@gerhard @gmile @helderco The PR is ready to review now. 🎉

File: sdk/elixir/runtime/dagger.json
Comment (wingyplus) [inline sdk/elixir/runtime/dagger.json:2]: I see typescript use capital case but in Python use kebab case. I'm not sure which one we should use but I follow Python SDK. 😂

Comment (helderco) [reply sdk/elixir/runtime/dagger.json:2]: Good point, I think kebab is the way to go, will see that it's normalized 🙂

Comment (wingyplus) [reply sdk/elixir/runtime/dagger.json:2]: Yeah. But I need to test more on this case because currently, I just normalize this name to lower case and replace - to _ to use in Elixir package name. :(

Comment (wingyplus) [reply sdk/elixir/runtime/dagger.json:2]: Fixed with `strcase`. :)

File: core/integration/module_elixir_test.go
Comment (wingyplus) [inline core/integration/module_elixir_test.go:35]: When I remove `-vv` I got 

```
=== NAME  TestModuleElixirInit/camel-cases_Dagger_module_name
    module_elixir_test.go:39:
                Error Trace:    /home/wingyplus/src/github.com/dagger/dagger/core/integration/module_elixir_test.go:39
                Error:          []string{".gitattributes", ".gitignore", "pyproject.toml", "requirements.lock", "sdk", "src"} does not contain "my_module"
                Test:           TestModuleElixirInit/camel-cases_Dagger_module_name
=== NAME  TestModuleElixirInit/with_source
    module_elixir_test.go:65:
                Error Trace:    /home/wingyplus/src/github.com/dagger/dagger/core/integration/module_elixir_test.go:65
                Error:          []string{".gitattributes", ".gitignore", "pyproject.toml", "requirements.lock", "sdk", "src"} does not contain "bare"
                Test:           TestModuleElixirInit/with_source
=== NAME  TestModuleElixirInit/from_scratch
```

Maybe because of I accidentally copy the python sdk at commit https://github.com/dagger/dagger/pull/6967/commits/e3f97af57a7ebfbec4b53420e5a4a2959f25949f. But the issue still occurred even correct the path. After try adding a flag, it works, change the directory name also works! 😢

**Comment by gmile** [comment]
For what it's worth, I tested this a few times now, all is working as expected, no complains from here. The biggest thing I was waiting is `terminal` sub-command, and it works like a charm. HUGE thank you for all the hard work, @wingyplus! 🥇👏 🎉

File: sdk/elixir/runtime/main.go
Comment (helderco) [inline sdk/elixir/runtime/main.go:16]: Best if full image ref, including the digest as well.

File: sdk/elixir/runtime/main.go
Comment (helderco) [inline sdk/elixir/runtime/main.go:73]: Shouldn't this be `genDir` instead?

File: sdk/elixir/runtime/main.go
Comment (helderco) [inline sdk/elixir/runtime/main.go:105]: This would be `genDir`

File: sdk/elixir/runtime/main.go
Comment (helderco) [inline sdk/elixir/runtime/main.go:143]: Don't see a need for acepting the `version` as function argument.

File: sdk/elixir/runtime/main.go
Comment (helderco) [inline sdk/elixir/runtime/main.go:148]: Consider just using a full image ref instead, saved as a const at top.

File: sdk/elixir/runtime/main.go
Comment (helderco) [inline sdk/elixir/runtime/main.go:96]: Consider order in pipeline steps as the sdkSrc won't change between releases while `modSource` depends on each module users run. So things that change less frequently should happen sooner to avoid cache invalidation.

File: sdk/elixir/runtime/main.go
Comment (helderco) [inline sdk/elixir/runtime/main.go:92]: There's no need to pass `dag` as it's a globally available client instance. See https://dagger.io/blog/simpler-reusability-with-the-global-client

You can see it in this module's `dagger.gen.go` file.

File: sdk/elixir/runtime/main.go
Comment (helderco) [inline sdk/elixir/runtime/main.go:96]: You don't seem to need the sdk in `/src`. Just put it in `genDir` (`<context directory> + <source subpath> + `sdk/`). You already seem to be doing it below but using `dagger` instead of `sdk`.

File: sdk/elixir/runtime/main.go
Comment (helderco) [inline sdk/elixir/runtime/main.go:117]: Since you're already in subpath at this point, you can do just `WithWorkdir("dagger")`. Doesn't have to be an absolute path.

File: sdk/elixir/runtime/main.go
Comment (helderco) [inline sdk/elixir/runtime/main.go:114]: Are all of these necessary? Consider excluding what you don't need from the build that puts these files in the engine container, and maybe do some of the steps above like a multi-stage build, so do you don't inherit everything. For example, you can have a separate pipeline for the codegen that just returns the generated files in the end, which you then add here. You won't have any other files from installing codegen, not need to change workdir back, etc, that you don't need afterwards.

File: sdk/elixir/runtime/main.go
Comment (helderco) [inline sdk/elixir/runtime/main.go:125]: If this takes more than ms, it can be skipped.

File: sdk/elixir/runtime/main.go
Comment (helderco) [inline sdk/elixir/runtime/main.go:129]: The pipeline could have any number of errors, this isn't a good assumption to have because those errors are being ignored here. Beter to check if the file exists with `Entries()`.

File: sdk/elixir/runtime/template.exs
Comment (helderco) [inline sdk/elixir/runtime/template.exs:59]: Please include all the docs the other SDKs use. We have a convention that the code any SDK generates produces the same result (in the API and CLI).

**Review by helderco — COMMENTED** [review]
Sorry for taking such a long time, I'm not at all familiar with Elixir. Took most notes on the runtime module. Since it's an experimental SDK, we can do more incremental improvements so my comments aren't merge blockers.

There's several tests in `core/integration/module_test.go` that put several SDKs under the same test, you just have to add the equivalent elixir code as a test case. For example:

https://github.com/dagger/dagger/blob/379e812dee7c36713c30040324b99559bd74cf36/core/integration/module_test.go#L547

https://github.com/dagger/dagger/blob/379e812dee7c36713c30040324b99559bd74cf36/core/integration/module_test.go#L1389

https://github.com/dagger/dagger/blob/379e812dee7c36713c30040324b99559bd74cf36/core/integration/module_test.go#L1733

See how many you can add. It's an easy way to check for consistency and add test coverage.

File: sdk/elixir/runtime/main.go
Comment (helderco) [reply sdk/elixir/runtime/main.go:129]: This is a required change for merge.

**Comment by wingyplus** [comment]
Thanks for reviewing @helderco. That was a huge comments! I'll take a look during this weekend. :)

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go:143]: You're right. I think we can remove it for now.

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go:117]: I see. Let me try.

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go:114]: Actually. I just copy the SDK source from `sdk/elixir` to under sub path and exclude unwanted files here. The reason is I wanted to keep SDK up-to-date without worry about dependencies or source code that's not covered by the codegen, such as graphql core, querybuilder, etc. When you generate by `dagger init` you will got 2 directories under `dagger` directory.

```
wingyplus@WINGYMOMO ~/s/g/w/d/potato (main)> ls -lah dagger
total 24K
drwxr-xr-x 4 wingyplus wingyplus 4.0K Apr 27 15:56 ./
drwxr-xr-x 3 wingyplus wingyplus 4.0K Apr 27 15:56 ../
-rw------- 1 wingyplus wingyplus   27 Apr 27 15:56 .gitattributes
-rw------- 1 wingyplus wingyplus    8 Apr 27 15:56 .gitignore
drwxr-xr-x 3 wingyplus wingyplus 4.0K Apr 27 15:55 dagger/
drwxr-xr-x 4 wingyplus wingyplus 4.0K Apr 27 15:56 potato/
```

The `dagger` is the SDK and the `potato` is the library from module name that added `dagger` as a dependencies.

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go:125]: 1s exactly. T_T

Screenshot 2024-04-21 164344

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go:92]: That's my habit that in Elixir there's no global state. 😂 Let me fix.

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go]: Fixed in https://github.com/dagger/dagger/pull/6967/commits/a644d96ed3a67dbc2dac46ad64bfd39bcaa32acf

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go]: Fixed in https://github.com/dagger/dagger/pull/6967/commits/a644d96ed3a67dbc2dac46ad64bfd39bcaa32acf

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go]: Fixed in https://github.com/dagger/dagger/pull/6967/commits/a644d96ed3a67dbc2dac46ad64bfd39bcaa32acf

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go]: Fixed  in https://github.com/dagger/dagger/pull/6967/commits/a644d96ed3a67dbc2dac46ad64bfd39bcaa32acf

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go:116]: Or I should rename directory to `sdk` to avoid confusion?

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go]: Fixed in https://github.com/dagger/dagger/pull/6967/commits/1764f20086747660408ced5ef45b8b9db65b242e

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go:93]: Fixed in https://github.com/dagger/dagger/pull/6967/commits/1764f20086747660408ced5ef45b8b9db65b242e

Comment (wingyplus) [reply sdk/elixir/runtime/main.go]: Fixed in https://github.com/dagger/dagger/pull/6967/commits/1764f20086747660408ced5ef45b8b9db65b242e

File: sdk/elixir/runtime/main.go
Comment (helderco) [reply sdk/elixir/runtime/main.go]: > That's my habit that in Elixir there's no global variable. 😂 Let me fix.

Just curious but is that because of how Elixir works or just not implemented yet in the Elixir SDK? 🙂

File: sdk/elixir/runtime/main.go
Comment (helderco) [reply sdk/elixir/runtime/main.go:110]: Yeah, I assumed the output after `dagger init` was correct. I'm talking about the pipeline, potentially simplifying it and avoiding cache invalidations. In this specific instance, you have a bunch of excludes when adding `m.SDKSourceDir`, but that directory comes from what you include in `ci/build/sdk.go` so it should already be "clean". As for the multi-stage, see how it's being done in python:

https://github.com/dagger/dagger/blob/379e812dee7c36713c30040324b99559bd74cf36/sdk/python/runtime/main.go#L268-L280

So those steps are separate from the main pipeline, which helps avoid cache invalidation because all of this produces a single input which is the generated/codegened file. Even if this sub-pipeline invalidates due to some temporary file during codegen installation for example, as long as the output is the same, it won't invalidate the main pipeline.

Just a suggestion, not a requirement 🙂

Comment (helderco) [reply sdk/elixir/runtime/main.go:110]: > Or I should rename directory to `sdk` to avoid confusion?

The default right now is for `dagger init`to put the sources in a `dagger` sub-directory. So this puts the SDK in a `dagger/dagger` directory. On the other hand, I've add trouble filtering out `sdk` from the root of our repo because we have an `sdk` directory there. Maybe `dagger_sdk`? 🤷‍♂️

Comment (wingyplus) [reply sdk/elixir/runtime/main.go:110]: > Yeah, I assumed the output after `dagger init` was correct. I'm talking about the pipeline, potentially simplifying it and avoiding cache invalidations. In this specific instance, you have a bunch of excludes when adding `m.SDKSourceDir`, but that directory comes from what you include in `ci/build/sdk.go` so it should already be "clean". As for the multi-stage, see how it's being done in python:
> 
> https://github.com/dagger/dagger/blob/379e812dee7c36713c30040324b99559bd74cf36/sdk/python/runtime/main.go#L268-L280
> 
> So those steps are separate from the main pipeline, which helps avoid cache invalidation because all of this produces a single input which is the generated/codegened file. Even if this sub-pipeline invalidates due to some temporary file during codegen installation for example, as long as the output is the same, it won't invalidate the main pipeline.
> 
> Just a suggestion, not a requirement 🙂

I see. I forgot that I just do filtering during packing sdk in ci/sdk/build.go. So we could delete this exclusion here to make it single source of truth for the exclusion list. 👍 

Back to the pipeline, I think I get it now after read the python source but need to play with it some more.

Comment (wingyplus) [reply sdk/elixir/runtime/main.go:110]: > > Or I should rename directory to `sdk` to avoid confusion?
> 
> The default right now is for `dagger init`to put the sources in a `dagger` sub-directory. So this puts the SDK in a `dagger/dagger` directory. On the other hand, I've add trouble filtering out `sdk` from the root of our repo because we have an `sdk` directory there. Maybe `dagger_sdk`? 🤷‍♂️

I think that might be a good name for it. 👍

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go]: > > That's my habit that in Elixir there's no global variable. 😂 Let me fix.
> 
> Just curious but is that because of how Elixir works or just not implemented yet in the Elixir SDK? 🙂

It's by design. If we want something like Go does, we need to create a new Process, a lightweight thread in Erlang/Elixir, like a goroutine and maintain a client inside it, maybe it need to clone the API interface to talk with the process. That's looks complex.

File: sdk/elixir/runtime/dagger.json
Comment (wingyplus) [inline sdk/elixir/runtime/dagger.json:5]: Is it ok to use this engine version? I can't find what version we should use  but it's always get generated by `dagger develop`.

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go:73]: I just ignore the `dagger` sdk path here. What it's incorrect might be `WithVCSGeneratedPaths`. But if we call `WithVCSIgnoredPaths` with this directory, we don't need to call `WithVCSGeneratedPaths`?

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go]: Fixed in  https://github.com/dagger/dagger/pull/6967/commits/43b8e316cd5b4ec77af2b97512a4cc0b35ed4c70

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go]: Improved in https://github.com/dagger/dagger/pull/6967/commits/b56617b7f36897fb773bdec543f69a645e677be1 and rename the sdk to `dagger_sdk` in the latest changes. 🙇

File: sdk/elixir/runtime/dagger.json
Comment (helderco) [reply sdk/elixir/runtime/dagger.json:5]: Yeah, it's ok.

File: sdk/elixir/runtime/main.go
Comment (helderco) [reply sdk/elixir/runtime/main.go]: Both of them should point to any generated files, including the SDK.

File: sdk/elixir/runtime/template.exs
Comment (wingyplus) [reply sdk/elixir/runtime/template.exs]: Fixed in https://github.com/dagger/dagger/pull/6967/commits/38bedae5a016b76304f1db492bb3825a6a0e8506.

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go]: I temporary removing it for now to make module startup faster.

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go]: Fixed.

File: sdk/elixir/runtime/main.go
Comment (wingyplus) [reply sdk/elixir/runtime/main.go]: Just set to genDir, which's `dagger_sdk`. (fixed in https://github.com/dagger/dagger/pull/6967/commits/b56617b7f36897fb773bdec543f69a645e677be1)

**Comment by wingyplus** [comment]
@helderco Unfortunately, I can do only one integration tests because most of it required to have struct supported, which's not covered by this PR. And found test `TestModuleDescription` is not played well with the current approach because Elixir `mix new` will prompt if it detect any file(s) in the directory that needs to be created.

**Review by helderco — CHANGES_REQUESTED** [review]
@wingyplus, I'm sorry to have taken so long to get back to you! There's one significant change I need you to make and I'll come back and revisit after you're done. 🙏

File: core/schema/sdk.go
Comment (wingyplus) [reply core/schema/sdk.go:95]: > let's not bundle community maintained SDKs in the engine container just yet. We may revisit at a later date, but for now users should use the git path to the runtime container in `--sdk`.

Agreed. Testing the runtime is quite slow because I need to do build the engine & cli while external runtime experience is much better. So I'll revert the change. 👍 



> There's been some ideas floating around for improving this, for example, by reserving a shorthand prefix (instead of requiring the full git ref) in order to be easier to type and fetch the sources at a standard location by URL convention, but the design for this isn't yet flushed out.

I have been thinking about having a sdk & runtime under dagger organization rather than having under dagger repository. This would make sdk maintainer iterate more faster since they have more control on that repository. But the release process would be more complex.

Comment (wingyplus) [reply core/schema/sdk.go:95]: Do we need to remove elixir test out of integration test or need to switch to git path?

Comment (helderco) [reply core/schema/sdk.go]: > I have been thinking about having a sdk & runtime under dagger organization rather than having under dagger repository. This would make sdk maintainer iterate more faster since they have more control on that repository. But the release process would be more complex.

Yeah, this has been suggested before (cc @gerhard) for that exact reason (also peformance would be better for the runtime module when loading for the first time), and same downside too.

There's also a balance in terms of control over the design. On one hand it allows faster iteration due to control, but on the other hand, having the team review changes helps in keeping alignment. The main problems are low bandwidth and probably lack of familiarity with the language.

Comment (helderco) [reply core/schema/sdk.go]: > Do we need to remove elixir test out of integration test or need to switch to git path? My opinion is temporary moving it out for now would be better. Adding to integration tests is quite a bit painful, especially finding log in the CI. 😂
> 
> However, I'm thinking about having a conformance test, like https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md. I see that in integration tests already have this kind of test but didn't expose the spec outside.

Yeah, I expect we'll want to have an easier system for integration tests for multiple SDKs, where you just write the code for modules that have to conform to certain expectations.

We'll feel the pressure for that more and more, as we need to test all the SDKs and more may be coming. \cc @jedevc

As for moving the tests out, yes if it's easier for you. Python and TypeScript have their own test suite in their own `sdk/<name>` dirs. And both have a `dev` module, written in their language, to help manage the SDK.

The idea is that you can call the dev module manually, but also add it as a dependency to the root's ci module, which then simply calls the dev module's functions rather than all the code being there.

It also helps with observability and concurrency when done like this.

Comment (wingyplus) [reply core/schema/sdk.go]: > As for moving the tests out, yes if it's easier for you. Python and TypeScript have their own test suite in their own `sdk/<name>` dirs. And both have a `dev` module, written in their language, to help manage the SDK.

I believe `dev` module is a good dogfood for the SDK itself. I'll go in that way. 👍

**Comment by wingyplus** [comment]
It seems testrace failed in `module_python_test.go`. Not sure why, so re-running it. :)

File: core/schema/sdk.go
Comment (wingyplus) [reply core/schema/sdk.go]: @helderco I think it's done now. Let me know if you have something to change. 🙇

**Comment by wingyplus** [comment]
@helderco ping. :)

**Comment by wingyplus** [comment]
Rebased.

**Comment by helderco** [comment]
> * Field support (like `@field` in TypeScript)

Don't use "field", there's an issue to rename TypeScript and Python's "field" to "function": _everything is a function_.

**Comment by helderco** [comment]
@wingyplus, can you update instructions in the description now that it's not a built-in module?

**Comment by helderco** [comment]
Hmmm... how do you pass the sdk's files when not a built-in runtime module? Have you been able to make it work?

**Comment by wingyplus** [comment]
> @wingyplus, can you update instructions in the description now that it's not a built-in module?

Sure!

**Comment by wingyplus** [comment]
> Hmmm... how do you pass the sdk's files when not a built-in runtime module? Have you been able to make it work?

Oh. I may missing something, let me check.

**Comment by wingyplus** [comment]
> Hmmm... how do you pass the sdk's files when not a built-in runtime module? Have you been able to make it work?

@helderco That's my mistake 🙏 . Fixed by clone it in `New` function instead but the url still invalid because I used `github.com/dagger/dagger` on branch `main` which's currently not available at the moment, we need to test after merge this PR. 

If you have any suggestion, feel free to comment.

**Comment by wingyplus** [comment]
> @wingyplus, can you update instructions in the description now that it's not a built-in module?

Done. Also added to the README.md but it's need a little polish after merge.

**Comment by helderco** [comment]
> [...] I used `github.com/dagger/dagger` on branch `main` which's currently not available at the moment, we need to test after merge this PR.

PRs have their own URL: https://github.com/dagger/dagger#pull/6967/head

See https://play.dagger.cloud/playground/COGeM4i4Z36

**Comment by wingyplus** [comment]
@helderco sorry for response to you lately. Here is the test result.

Screenshot 2024-05-26 155530

I do test on commit https://github.com/dagger/dagger/pull/6967/commits/67bb3c7b83f98893b07a83cb81720362fefd1429 and switch back to dagger main branch on https://github.com/dagger/dagger/pull/6967/commits/2a918f80f93d7cfe38a6ee943ffbf6803ea24851

**Review by helderco — APPROVED** [review]
All good 👍

Can you rebase and see if checks go green?

**Comment by wingyplus** [comment]
@helderco Rebased. Waiting for checks. 🙏

**Comment by helderco** [comment]
Nice work @wingyplus 🎉

**Comment by wingyplus** [comment]
Thank you @helderco 🎉🎉

**Comment by gmile** [comment]
Awesome work @wingyplus and @helderco!!! Happy to see this finally land 👏

**Comment by gerhard** [comment]
💚 @wingyplus

**Pull request merged**
